clang/lib/CIR/Dialect/IR/CIRDialect.cpp

```diff
@@ -3854,7 +3854,9 @@ LogicalResult cir::StdInitializerListOp::verify() {
     return emitOpError("first member type of std::initializer_list must be "
                        "'!cir.ptr', but provided ")
            << resultType.getMembers()[0];
-  auto expectedType = memberPtr.getPointee();
+  mlir::Type expectedType = memberPtr.getPointee();
+  if (mlir::dyn_cast<cir::StructType>(expectedType))
```

Comment:
Better be careful here. What about array type?

Reply:
Is there any case to we can create std::initializer_list with array type?

It does not work
```c++
void f(std::initializer_list<int[2]>);
void test() {
  int a[2] = {1,2};
  f({a,a,a});
}
```

Reply:
Thanks for fixing it. Array type is an apparent one of the cases. I don't know how even complex types (e.g. those introduced by exceptions) should work, and what we should do to maintain the code when extending the type system of CIR. Options could be like, collecting predicates in one single place or adding a FIXME here.

Sorry for blocking it. The changes may not necessarily be included in this patch. You can wait @bcardosolopes for his advice ; )

Reply:
I also fell uncomfortable for this code. The reason is in CIR, we actually have 2 kinds of value, one pass by value, the other pass by pointer. Could we somehow align them?
the full support for `cir.std.initializer_list` is still on going. And I am confused how to handle this two kinds. Especially for the struct with deconstruct function. The lifetime of temporary variable will be hard to manage.

Reply:
IMHO, this is another orthogonal design choice to retain some low-level constructs, such as pointers, in function calls. It's worth noting that CIR is currently somewhat low-level by design, as gradual improvements are planned. There are also many discussion on it in progress, e.g. [this one](https://discourse.llvm.org/t/rfc-a-clangir-based-safe-c/83245#p-334860-what-is-the-status-quo-of-clangir-for-analysis-31) that you may be interested in ; )

In the last patch for `cir.std.init_list`, I believe that we essentially agreed on treating this op as something akin to a "macro" that organizes inputs more effectively while maintaining a clear correspondence to the original low-level ops (introduced in #764). Do you think such approach has any drawbacks?

Reply:
> IMHO, this is another orthogonal design choice to retain some low-level constructs, such as pointers, in function calls. It's worth noting that CIR is currently somewhat low-level by design, as gradual improvements are planned. There are also many discussion on it in progress, e.g. [this one](https://discourse.llvm.org/t/rfc-a-clangir-based-safe-c/83245#p-334860-what-is-the-status-quo-of-clangir-for-analysis-31) that you may be interested in ; )

Maybe we need a clearer guideline to point which type will pass by ptr. e.g. most of struct is passed by pointer. But `cir.std.initializer_list` will create a struct directly. What will happen if we want to construct a initializer_list of initializer_list?

Reply:
sorry, i confused by some outdated design which contains alloca op in std.initlist also.

Reply:
If we distinguish cir type with 2 kinds. one is by ptr, the other is by value. the current design IMO is fine. it is not so harmful to support two kinds of input.

for value type, we must store it to array. it is the same as current llvmir.
for struct/array, construct a tmp rvalue and move it to array IMO also fulfill standard and can be optimized later.
it still remains some questions, e.g. when to deconstruct tmp value. i am trying to find a solution now.

Reply:
Sorry for the late reply. Last week was Thanksgiving week in the US, and I fell behind afterwards.

This is an interesting case that I didn't think about at all during the original review; thanks for noticing it and putting up the PR. I agree that the fundamental fix would be having value semantics. Doing the value vs. pointer type split might work for now, but it feels fragile. I know @bcardosolopes had some thoughts here too; he's traveling right now but he can weigh in when he gets the chance.

Slightly unrelated, but I'm working on https://github.com/compiler-explorer/clang-builder/issues/75 to get us a more up-to-date ClangIR compiler on Godbolt.

Reply:
@HerrCai0907 I was giving this more thought, and I think we should take a step back:
- We should revert the PR that introduced cir.std.initializer_list. 
- Introduce a new PR where cir.std.initializer_list is just a thin layer over the initialization work, for example:
```
%0 = cir.alloca INITLIST_TYPE, !cir.ptr<INITLIST_TYPE>
cir.std.initializer_list {
  %1 = cir.const #cir.int<1>
  ... // all init work goes here
}
```
My rationale is that we don't yet have a pass taking advantage of this and we should probably use that to guide how we really want to design this, anything before that will be more or less a guess game. Once we have usage going through it and we start writing a pass that wants to look at the initialization data, then we refine. Sorry for the churn, WDYT?